File: paleo_workbench/ui/pages/data_workspace.py
from __future__ import annotations

from PySide6.QtCore import QEvent, Qt, QTimer
from PySide6.QtWidgets import (
    QHBoxLayout,
    QSplitter,
    QStackedWidget,
    QToolButton,
    QVBoxLayout,
    QWidget,
)

from paleo_workbench.ui.dock_manager import dock_manager
from paleo_workbench.ui.layout_persistence import LayoutPersistence
from paleo_workbench.ui.panel_float_controller import FloatController
from paleo_workbench.ui.pages.data_asset_table import DataAssetTable
from paleo_workbench.ui.pages.data_reader_panel import DataReaderPanel
from paleo_workbench.ui.pages.inspector_panel import InspectorPanel
from paleo_workbench.ui.pages.navigation_tree import NavigationTree
from paleo_workbench.ui.pages.project_overview_panel import ProjectOverviewPanel
from paleo_workbench.ui.pages.well_map_panel import WellMapPanel

#: Delay before a splitter drag is persisted (avoid a QSettings sync per tick).
_DOCKED_SIZES_DELAY_MS = 400


class PanelFloatButton(QToolButton):
    """Corner button floating one side panel via a FloatController.

    Docked side of the FloatingPanel ⇲ dock-back chrome: pinned to the
    panel's top-right corner through an event filter and hidden while the
    panel is afloat (the floating window carries its own dock-back button).
    Pass ``pin=False`` to place the button into an existing header strip
    (the well-map fold header) instead of overlaying the panel corner.
    """

    def __init__(
        self,
        key: str,
        panel: QWidget,
        controller: FloatController,
        *,
        pin: bool = True,
    ):
        super().__init__(panel if pin else None)
        self._key = key
        self._panel = panel
        self._controller = controller
        self._pinned = pin
        self.setObjectName("PanelFloatButton")
        self.setText("⇱")
        self.setToolTip("浮动面板 (Float panel)")
        self.setFixedSize(18, 18)
        self.clicked.connect(lambda: self._controller.toggle(self._key))
        if pin:
            panel.installEventFilter(self)
        controller.float_changed.connect(self._on_float_changed)
        if pin:
            self._reposition()

    def _reposition(self) -> None:
        self.move(self._panel.width() - self.width() - 4, 2)

    def eventFilter(self, obj, event) -> bool:  # noqa: N802 (Qt override)
        if obj is self._panel and event.type() == QEvent.Type.Resize:
            self._reposition()
        return False

    def _on_float_changed(self, key: str, floating: bool) -> None:
        if key == self._key:
            self.setVisible(not floating)


class DataWorkspace(QWidget):
    def __init__(
        self,
        parent=None,
        *,
        well_state_store=None,
        comparison_crs: str = "",
        persistence: LayoutPersistence | None = None,
    ):
        super().__init__(parent)
        self.setObjectName("DataWorkspace")

        layout = QHBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(0)

        self.main_splitter = QSplitter(Qt.Orientation.Horizontal)
        self.main_splitter.setObjectName("DataMainSplitter")
        self.main_splitter.setChildrenCollapsible(False)

        self.navigation_tree = NavigationTree()
        self.asset_table = DataAssetTable()

        # Center column: 工区概览 panel stacked over the asset table.  The
        # 工区概览 tree node swaps the stack; every other node shows the table.
        self._center_stack = QStackedWidget()
        self._center_stack.setObjectName("DataCenterStack")
        self.overview_panel = ProjectOverviewPanel()
        self._center_stack.addWidget(self.asset_table)  # index 0 = table
        self._center_stack.addWidget(self.overview_panel)  # index 1 = overview

        # Right column: vertical splitter of reader + inspector
        self.right_splitter = QSplitter(Qt.Orientation.Vertical)
        self.right_splitter.setChildrenCollapsible(False)
        self.reader_panel = DataReaderPanel(
            well_state_store=well_state_store,
            comparison_crs=comparison_crs,
        )
        self.inspector_panel = InspectorPanel()
        self.right_splitter.addWidget(self.reader_panel)
        self.right_splitter.addWidget(self.inspector_panel)
        self.right_splitter.setStretchFactor(0, 2)
        self.right_splitter.setStretchFactor(1, 1)
        self.right_splitter.setSizes([400, 200])

        self.main_splitter.addWidget(self.navigation_tree)
        center_container = QWidget()
        self._map_center_host = center_container
        self._center_layout = QVBoxLayout(center_container)
        self._center_layout.setContentsMargins(0, 0, 0, 0)
        self._center_layout.addWidget(self._center_stack, 1)
        # 井位地图 collapsible panel under the table (§18: absorbed page).
        # While the 工区概览 node is active it moves into the overview page
        # as the main content (see show_overview).
        self.well_map_panel = WellMapPanel()
        self._center_layout.addWidget(self.well_map_panel, 0)
        self._map_collapsed_before_overview = True
        self.main_splitter.addWidget(center_container)
        self.main_splitter.addWidget(self.right_splitter)
        self.main_splitter.setStretchFactor(0, 0)
        self.main_splitter.setStretchFactor(1, 1)
        self.main_splitter.setStretchFactor(2, 0)
        self.main_splitter.setSizes([220, 600, 480])

        layout.addWidget(self.main_splitter)

        # M6: the side panels float through the shared FloatController; the
        # center table/overview stack stays docked.
        self._float_persistence = (
            persistence if persistence is not None else LayoutPersistence()
        )
        self._floatable: dict[str, QWidget] = {}
        self.float_controller = FloatController(
            resolver=self._floatable.get,
            persistence=self._float_persistence,
            parent=self,
        )
        self._make_floatable("data:navigation", self.navigation_tree, "数据导航树")
        self._make_floatable("data:reader", self.reader_panel, "数据预览")
        self._make_floatable("data:inspector", self.inspector_panel, "数据资产检查器")
        self._make_floatable("data:well_map", self.well_map_panel, "井位地图")
        # 井位地图 has its own fold header; its float toggle lives there
        # instead of the corner overlay.
        self.well_map_panel.add_header_button(
            PanelFloatButton(
                "data:well_map", self.well_map_panel, self.float_controller, pin=False
            )
        )
        self.float_controller.float_changed.connect(self._on_map_float_changed)
        self._map_collapsed_before_float = True
        self._float_sizes_timer = QTimer(self)
        self._float_sizes_timer.setSingleShot(True)
        self._float_sizes_timer.setInterval(_DOCKED_SIZES_DELAY_MS)
        self._float_sizes_timer.timeout.connect(self._persist_docked_sizes)
        self.main_splitter.splitterMoved.connect(self._on_splitter_moved)
        self.right_splitter.splitterMoved.connect(self._on_splitter_moved)
        for key in self._floatable:
            self.float_controller.restore_saved(key)

    def _make_floatable(self, key: str, panel: QWidget, title: str) -> None:
        """Register a side panel for float/dock and give it a float button."""
        dock_manager.register_panel(key, title)
        self._floatable[key] = panel
        PanelFloatButton(key, panel, self.float_controller)

    def _on_splitter_moved(self, *_pos: int) -> None:
        self._float_sizes_timer.start()

    def _on_map_float_changed(self, key: str, floating: bool) -> None:
        """Expand the afloat map and rebuild its fold slot after docking.

        FloatController re-parents to the recorded dock context, but a
        layout-parented widget needs an explicit layout re-insert to take
        part in the layout again (its splitter restore can also drop it into
        an ancestor splitter slot); a collapsed map would surface as an empty
        floating window, so floating expands and docking restores the fold.
        """
        if key != "data:well_map":
            return
        if floating:
            self._map_collapsed_before_float = self.well_map_panel.is_collapsed()
            self.well_map_panel.set_collapsed(False)
            return
        lost_from_fold = (
            self.well_map_panel.parentWidget() is not self._map_center_host
            or self._center_layout.indexOf(self.well_map_panel) == -1
        )
        if lost_from_fold:
            self._center_layout.addWidget(self.well_map_panel, 0)
        self.well_map_panel.set_collapsed(self._map_collapsed_before_float)

    def _persist_docked_sizes(self) -> None:
        """Persist the splitter distribution for every docked side panel."""
        for key, panel in self._floatable.items():
            splitter = panel.parentWidget()
            if isinstance(splitter, QSplitter):
                self._float_persistence.save_docked_sizes(key, list(splitter.sizes()))

    def show_overview(self, visible: bool) -> None:
        """工区概览 node ↔ table swap (center column only).

        Overview mode reparents the shared well-map panel into the overview
        page — 工区概况下面直接显示工区图，展开占满主区域；切回数据表时
        地图回到表格下方的折叠面板并恢复之前的折叠状态。
        While the map panel is floating the swap is skipped entirely: the
        controller owns the afloat widget until it is docked back.
        """
        map_afloat = self.float_controller.is_floating("data:well_map")
        self._center_stack.setCurrentIndex(1 if visible else 0)
        if visible:
            self._map_collapsed_before_overview = self.well_map_panel.is_collapsed()
            if map_afloat:
                return
            self.overview_panel.set_map_widget(self.well_map_panel)
            self.well_map_panel.set_header_visible(False)
            self.well_map_panel.set_collapsed(False)
        else:
            if map_afloat:
                return
            self._center_layout.addWidget(self.well_map_panel, 0)
            self.well_map_panel.set_header_visible(True)
            self.well_map_panel.set_collapsed(self._map_collapsed_before_overview)

    def overview_visible(self) -> bool:
        return self._center_stack.currentIndex() == 1

    def set_right_visible(self, visible: bool) -> None:
        self.right_splitter.setVisible(visible)

File: libs/ui_pages_data/src/qt/data_workspace.cpp
// UI-06 — DataWorkspace shell (see qt/data_workspace.hpp).
#include <pwb/ui_pages_data/qt/data_workspace.hpp>

#include <QEvent>
#include <QHBoxLayout>
#include <QSizePolicy>
#include <QTimer>
#include <QVBoxLayout>

#include <pwb/ui_pages_data/qt/asset_selection_bus.hpp>
#include <pwb/ui_pages_data/qt/data_asset_table.hpp>
#include <pwb/ui_pages_data/qt/data_reader_panel.hpp>
#include <pwb/ui_pages_data/qt/navigation_tree_widget.hpp>
#include <pwb/ui_pages_data/qt/project_overview_panel.hpp>

namespace pwb::ui_pages_data::qt {

// --- PanelFloatButton ---------------------------------------------------------

PanelFloatButton::PanelFloatButton(const QString& key, QWidget* panel,
                                   FloatControllerApi* controller,
                                   bool pin, QWidget* parent)
    : QToolButton(pin ? panel : parent),
      key_(key),
      panel_(panel),
      controller_(controller),
      pinned_(pin) {
    setObjectName(QStringLiteral("PanelFloatButton"));
    setText(QStringLiteral("⇱"));
    setToolTip(QStringLiteral("浮动面板 (Float panel)"));
    setFixedSize(18, 18);
    connect(this, &QToolButton::clicked, this, [this] {
        if (controller_ != nullptr) controller_->toggle(key_);
    });
    if (pinned_) panel_->installEventFilter(this);
    if (controller_ != nullptr) {
        connect(controller_, &FloatControllerApi::float_changed, this,
                [this](const QString& key, bool floating) {
                    if (key == key_) setVisible(!floating);
                });
    }
    if (pinned_) reposition();
}

void PanelFloatButton::reposition() {
    move(panel_->width() - width() - 4, 2);
}

bool PanelFloatButton::eventFilter(QObject* obj, QEvent* event) {
    if (obj == panel_ && event->type() == QEvent::Type::Resize) {
        reposition();
    }
    return false;
}

// --- DataWorkspace ------------------------------------------------------------

DataWorkspace::DataWorkspace(QWidget* parent) : QWidget(parent) {
    setObjectName(QStringLiteral("DataWorkspace"));
    auto* layout = new QHBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(0);

    main_splitter_ = new QSplitter(Qt::Orientation::Horizontal, this);
    main_splitter_->setObjectName(QStringLiteral("DataMainSplitter"));
    main_splitter_->setChildrenCollapsible(false);

    navigation_tree_ = new NavigationTree(this);
    asset_table_ = new DataAssetTable(this);

    center_stack_ = new QStackedWidget(this);
    center_stack_->setObjectName(QStringLiteral("DataCenterStack"));
    overview_panel_ = new ProjectOverviewPanel(this);
    well_detail_panel_ = new QWidget(this);  // seam placeholder
    center_stack_->addWidget(asset_table_);      // index 0 = table
    center_stack_->addWidget(overview_panel_);   // index 1 = overview
    center_stack_->addWidget(well_detail_panel_);// index 2 = well view

    // 原型 ws0 右列 = 数据属性（上）+ 数据血缘/处理流程（下）；两个都是
    // seam 槽位 —— V14 装配注入 DataDetailPanel / DataLineagePanel。
    right_splitter_ = new QSplitter(Qt::Orientation::Vertical, this);
    right_splitter_->setChildrenCollapsible(false);
    inspector_panel_ = new QWidget(this);  // 数据属性 seam placeholder
    lineage_panel_ = new QWidget(this);    // 数据血缘 seam placeholder
    right_splitter_->addWidget(inspector_panel_);
    right_splitter_->addWidget(lineage_panel_);
    right_splitter_->setStretchFactor(0, 3);
    right_splitter_->setStretchFactor(1, 2);
    right_splitter_->setSizes({260, 140});

    main_splitter_->addWidget(navigation_tree_);
    auto* center_container = new QWidget(this);
    map_center_host_ = center_container;
    center_layout_ = new QVBoxLayout(center_container);
    center_layout_->setContentsMargins(0, 0, 0, 0);
    // 原型 ws0 中央 = 数据列表 + 表格下页签（数据预览|版本历史|关联关系）。
    // 预览是真实 DataReaderPanel；版本/血缘页由 V14 装配提入。
    center_vsplit_ = new QSplitter(Qt::Orientation::Vertical,
                                   center_container);
    center_vsplit_->setObjectName(QStringLiteral("DataCenterSplit"));
    center_vsplit_->setChildrenCollapsible(false);
    center_vsplit_->addWidget(center_stack_);
    bottom_tabs_ = new QTabWidget(center_vsplit_);
    bottom_tabs_->setObjectName(QStringLiteral("DataBottomTabs"));
    reader_panel_ = new DataReaderPanel(bottom_tabs_);
    bottom_tabs_->addTab(reader_panel_, QStringLiteral("数据预览"));
    center_vsplit_->addWidget(bottom_tabs_);
    center_vsplit_->setStretchFactor(0, 11);
    center_vsplit_->setStretchFactor(1, 9);
    center_layout_->addWidget(center_vsplit_, 1);
    well_map_panel_ = nullptr;  // injected via set_well_map_panel
    main_splitter_->addWidget(center_container);
    main_splitter_->addWidget(right_splitter_);
    for (QWidget* side :
         {static_cast<QWidget*>(navigation_tree_), center_container,
          static_cast<QWidget*>(right_splitter_)}) {
        side->setSizePolicy(QSizePolicy::Policy::Expanding,
                            QSizePolicy::Policy::Expanding);
    }
    main_splitter_->setStretchFactor(0, 0);
    main_splitter_->setStretchFactor(1, 3);
    main_splitter_->setStretchFactor(2, 1);

    layout->addWidget(main_splitter_);

    float_sizes_timer_ = new QTimer(this);
    float_sizes_timer_->setSingleShot(true);
    float_sizes_timer_->setInterval(kDockedSizesDelayMs);
    connect(float_sizes_timer_, &QTimer::timeout, this,
            &DataWorkspace::persist_docked_sizes);
    connect(main_splitter_, &QSplitter::splitterMoved, this,
            [this](int, int) { float_sizes_timer_->start(); });
    connect(right_splitter_, &QSplitter::splitterMoved, this,
            [this](int, int) { float_sizes_timer_->start(); });
}

void DataWorkspace::showEvent(QShowEvent* event) {
    QWidget::showEvent(event);
    // 原型 ws0：数据列表约占中央高度 55%，表格下页签（数据预览|
    // 版本历史|关联关系）占余下 —— 首次可见时按真实高度播种一次；
    // 用户拖动后 persist_docked_sizes 接管，种子不再干预。
    if (vsplit_seeded_ || center_vsplit_ == nullptr) return;
    const int total = center_vsplit_->height();
    if (total <= 200) return;
    vsplit_seeded_ = true;
    const int top = total * 55 / 100;
    center_vsplit_->setSizes({top, total - top});
}

void DataWorkspace::set_float_controller(FloatControllerApi* controller) {
    float_controller_ = controller;
    if (float_controller_ == nullptr) return;
    make_floatable(QStringLiteral("data:navigation"), navigation_tree_,
                   QStringLiteral("数据导航树"));
    make_floatable(QStringLiteral("data:reader"), reader_panel_,
                   QStringLiteral("数据预览"));
    make_floatable(QStringLiteral("data:inspector"), inspector_panel_,
                   QStringLiteral("数据资产检查器"));
    if (well_map_panel_ != nullptr) {
        make_floatable(QStringLiteral("data:well_map"), well_map_panel_,
                       QStringLiteral("井位地图"));
        well_map_panel_->add_header_button(new PanelFloatButton(
            QStringLiteral("data:well_map"), well_map_panel_,
            float_controller_, /*pin=*/false));
    }
    connect(float_controller_, &FloatControllerApi::float_changed, this,
            &DataWorkspace::on_map_float_changed);
    for (const auto& [key, panel] : floatable_) {
        float_controller_->restore_saved(key);
    }
}

void DataWorkspace::set_save_docked_sizes_fn(
    std::function<void(const QString&, const QList<int>&)> fn) {
    save_docked_sizes_fn_ = std::move(fn);
}

void DataWorkspace::set_well_map_panel(WellMapPanelApi* panel) {
    if (well_map_panel_ != nullptr) {
        center_layout_->removeWidget(well_map_panel_);
    }
    well_map_panel_ = panel;
    if (well_map_panel_ != nullptr) {
        center_layout_->addWidget(well_map_panel_, 0);
    }
}

void DataWorkspace::set_inspector_panel(QWidget* panel) {
    if (panel == nullptr || panel == inspector_panel_) return;
    const int index = right_splitter_->indexOf(inspector_panel_);
    auto* old = inspector_panel_;
    inspector_panel_ = panel;
    // #1389: keep the floatable registry pointed at the LIVE panel — the
    // old one is deleteLater'd and a stale entry would dangle when the
    // docked-sizes debounce timer fires.
    auto fit = floatable_.find(QStringLiteral("data:inspector"));
    if (fit != floatable_.end()) {
        fit->second = panel;
    }
    if (index >= 0) {
        right_splitter_->insertWidget(index, panel);
        old->hide();
        old->deleteLater();
    }
}

void DataWorkspace::set_lineage_panel(QWidget* panel) {
    if (panel == nullptr || panel == lineage_panel_) return;
    const int index = right_splitter_->indexOf(lineage_panel_);
    auto* old = lineage_panel_;
    lineage_panel_ = panel;
    if (index >= 0) {
        right_splitter_->insertWidget(index, panel);
        old->hide();
        old->deleteLater();
    }
}

void DataWorkspace::set_well_detail_panel(QWidget* panel) {
    if (panel == nullptr || panel == well_detail_panel_) return;
    const int index = center_stack_->indexOf(well_detail_panel_);
    auto* old = well_detail_panel_;
    well_detail_panel_ = panel;
    if (index >= 0) {
        center_stack_->insertWidget(index, panel);
        center_stack_->removeWidget(old);
        old->deleteLater();
    }
}

void DataWorkspace::bind_selection_bus(AssetSelectionBus* bus) {
    if (selection_bus_ != nullptr) {
        disconnect(selection_bus_, nullptr, this, nullptr);
        // The table->bus lambda's receiver context is the bus itself —
        // drop it explicitly or an unbound table keeps feeding an orphan.
        disconnect(asset_table_, nullptr, selection_bus_, nullptr);
    }
    selection_bus_ = bus;
    if (selection_bus_ == nullptr) return;
    // Table → bus (the user's real selection is the state). The loop
    // bus→table→bus terminates on the bus's unchanged-selection check.
    connect(asset_table_, &DataAssetTable::selected_asset_changed,
            selection_bus_, [this](const std::optional<AssetRow>& asset) {
                if (syncing_selection_bus_) return;
                selection_bus_->set_current_asset(asset);
            });
    // Bus → table (external selection, row deletion, project switch).
    connect(selection_bus_,
            &AssetSelectionBus::current_asset_changed, this,
            [this](const std::optional<AssetRow>& asset) {
                syncing_selection_bus_ = true;
                asset_table_->set_selected_asset(asset.has_value()
                                                     ? &*asset
                                                     : nullptr);
                syncing_selection_bus_ = false;
            });
    connect(selection_bus_, &AssetSelectionBus::assets_changed, this,
            [this](const std::vector<AssetRow>& rows, const QString&) {
                syncing_selection_bus_ = true;
                asset_table_->update_assets(rows);
                syncing_selection_bus_ = false;
            });
}

void DataWorkspace::make_floatable(const QString& key, QWidget* panel,
                                   const QString& /*title*/) {
    floatable_[key] = panel;
    new PanelFloatButton(key, panel, float_controller_);
}

void DataWorkspace::on_map_float_changed(const QString& key,
                                         bool floating) {
    if (key != QLatin1String("data:well_map") ||
        well_map_panel_ == nullptr) {
        return;
    }
    if (floating) {
        map_collapsed_before_float_ = well_map_panel_->is_collapsed();
        well_map_panel_->set_collapsed(false);
        return;
    }
    const bool lost_from_fold =
        well_map_panel_->parentWidget() != map_center_host_ ||
        center_layout_->indexOf(well_map_panel_) == -1;
    if (lost_from_fold) {
        center_layout_->addWidget(well_map_panel_, 0);
    }
    well_map_panel_->set_collapsed(map_collapsed_before_float_);
}

void DataWorkspace::persist_docked_sizes() {
    if (!save_docked_sizes_fn_) return;
    for (const auto& [key, panel] : floatable_) {
        if (panel == nullptr) continue;  // destroyed between events (#1389)
        if (auto* splitter =
                qobject_cast<QSplitter*>(panel->parentWidget())) {
            save_docked_sizes_fn_(key, splitter->sizes());
        }
    }
}

void DataWorkspace::show_overview(bool visible) {
    const bool map_afloat =
        float_controller_ != nullptr &&
        float_controller_->is_floating(QStringLiteral("data:well_map"));
    center_stack_->setCurrentIndex(visible ? 1 : 0);
    if (well_map_panel_ == nullptr) return;
    if (visible) {
        map_collapsed_before_overview_ = well_map_panel_->is_collapsed();
        if (map_afloat) return;
        overview_panel_->set_map_widget(well_map_panel_);
        well_map_panel_->set_header_visible(false);
        well_map_panel_->set_collapsed(false);
    } else {
        if (map_afloat) return;
        center_layout_->addWidget(well_map_panel_, 0);
        well_map_panel_->set_header_visible(true);
        well_map_panel_->set_collapsed(map_collapsed_before_overview_);
    }
}

bool DataWorkspace::overview_visible() const {
    return center_stack_->currentIndex() == 1;
}

void DataWorkspace::show_well_detail(bool visible) {
    if (visible) {
        center_stack_->setCurrentIndex(2);
    } else if (center_stack_->currentIndex() == 2) {
        center_stack_->setCurrentIndex(0);
    }
}

bool DataWorkspace::well_detail_visible() const {
    return center_stack_->currentIndex() == 2;
}

void DataWorkspace::set_right_visible(bool visible) {
    right_splitter_->setVisible(visible);
}

}  // namespace pwb::ui_pages_data::qt
